test/integration/gzip_filter_integration_test.cc

```diff
@@ -0,0 +1,259 @@
+#include "common/decompressor/zlib_decompressor_impl.h"
+
+#include "test/integration/http_integration.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class GzipIntegrationTest : public HttpIntegrationTest,
+                            public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  GzipIntegrationTest() : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  void SetUp() override { decompressor_.init(window_bits); }
```

Comment:
these integration tests are great; sorry it's taking me a while to get through them.  But I didn't see one that I was expecting to see -- it might be there but I just didn't find it.

Have the simulated origin generate back some complex text (e.g. lorem ipsum...).  Check that the gzipped response is substantially smaller, and that uncompressing that response yields back the original content.

Reply:
What happens if the input text is not compressible; that is, the compressed output is larger than the uncompressed output.  This may occur if the origin content is already encoded with a same/better compressor (e.g. brotl or higher level of gzip), and it might be that the site wants the payload to stay that way and is not declaring it with a transfer-encoding. 

IMO the resultant response should not add another level of compression; it should just pass-through the headers and data unchanged.

Reply:
I assume that if the content is already compressed there will be a content-encoding header advertising it. I this case the filter will skip compression and leave the original content intact. 
```
if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
      !headers.ContentEncoding()) { ..
```
Unless the content is compressed and the content-encoding header is missing or misleading, but I think that's very unlike to happen right?

Reply:
If the content encoding is plan but the content itself was compressed (say, downloading a gzip file) then you'd have that scenario. One idiom I've seen before is to compare the compressed size to the original size after doing the compression and throw away the "compressed" result if it ends up bigger than the original, I think this is what @jmarantz is suggesting. This might best be an optimization that goes into a follow up PR, though.

Reply:
Yeah I think we can't do that unless we buffer.  We have to commit to the response headers prior to seeing any bytes of content to know whether they look compressible.  All we can go on is the content-type.